Replace redis-rb-cluster examples with official Valkey GLIDE client using original Ruby logic (#331)
valkey-io/valkey-doc: topics/cluster-tutorial.md

configuration, for example after a failover or after the system administrator
changed the cluster layout by adding or removing nodes.

#### Write an example app with Valkey GLIDE

Before going forward showing how to operate the Valkey Cluster, doing things
like a failover, or a resharding, we need to create some example application
or at least to be able to understand the semantics of a Valkey Cluster
client interaction.

In this way we can run an example and at the same time try to make nodes
failing, or start a resharding, to see how Valkey Cluster behaves under real
world conditions. It is not very helpful to see what happens while nobody
is writing to the cluster.

This section showcases the core functionality of Valkey through a practical Node.js application.  
For our example, we will use the [Node.js version of GLIDE,](https://github.com/valkey-io/valkey-glide/tree/main/node) an official Valkey client library that supports multiple languages.

The following example demonstrates how to connect to a Valkey cluster and perform
basic operations. First, install the Valkey GLIDE client:

```bash
npm install @valkey/valkey-glide
```

Here's the example code:

> ```javascript

Reviewer: @copilot imitate the same logic as the old implementation, just the equivalent in valkey-glide

                await client.mset(keyValuePairs);
                
                // Update counter and display progress
                await client.set("__last__", end.toString());
                console.log(`Batch completed: ${start} to ${end}`);
                
                // Verify a sample key from the batch
                const sampleKey = `foo${start}`;
                const value = await client.get(sampleKey);
                console.log(`Sample verification - ${sampleKey}: ${value}`);
                
            } catch (error) {
                console.log(`Error in batch starting at ${start}: ${error.message}`);
            }
        }
    } catch (error) {
        console.log(`Connection error: ${error.message}`);
    } finally {
        client.close();
    }
}

runExample().catch(console.error);
```

The application writes keys in the format `foo<number>` with their corresponding numeric values, using batched `MSET` operations for better performance. The `MSET` command accepts an array of alternating keys and values. So if you run the program the result is batches of `MSET` commands:

* `MSET foo1 1 foo2 2 foo3 3 ... foo100 100` (batch of 100 keys)

> * `MSET foo101 101 foo102 102 ... foo200 200` (next batch)

Reviewer: Glide handle multi-slot mget behind the scene, cluster-aware mset mget del flush

> * `MSET foo101 101 foo102 102 ... foo200 200` (next batch)

Reviewer 2: It's not atomic then. If it fails on one node, it's still set on the other nodes. This is risky and should not be the default behavior IMHO.

> * `MSET foo101 101 foo102 102 ... foo200 200` (next batch)

Reviewer 3: > Or update https://valkey.io/topics/command-tips/

Probably this one, just to clarify that it is not atomic anymore. MSET very clearly says it's atomic. 

> I always believe the code. But.... :) From my experience user asked for both atomic and non atomic MSET. Both make sense, we just need to be consistent.

I agree with this statement. I think users may want both. We should just be clear about what they are getting.

> * `MSET foo101 101 foo102 102 ... foo200 200` (next batch)

Reviewer 4: Two things:
1. I noted that the original `SET` example version doesn't actually do the same thing as the original. It just writes a bunch of keys were as the original wrote a bunch of keys then did a verification with `GET`. Additionally, the batching seems kind of... useless in terms of Valkey? It's just doing batches for Javascript, not in Valkey. I think the performance will be quite different:`SET` x 100s vs `MSET` x100 strings and both are non-atomic. Would it be better to do `MSET` with coordinated hash tags? (e.g. batches of 100 with a hash tag). I dunno - I'm generally not wild about this as an example for cluster.

2. @asafpamzn pointed out several points where the documentation is wrong. Can we get issues up for that so it doesn't get lost when we merge this PR?

> * `MSET foo101 101 foo102 102 ... foo200 200` (next batch)

Reviewer: @copilot open a follow up issue with @asafpamzn points

> * `MSET foo101 101 foo102 102 ... foo200 200` (next batch)

Reviewer: where?

is writing to the cluster.

This section explains some basic usage of
[Valkey GLIDE for Node.js](https://github.com/valkey-io/valkey-glide/tree/main/node), the official
Valkey client library, showing a simple example application.

The following example demonstrates how to connect to a Valkey cluster and perform
basic operations. First, install the Valkey GLIDE client:

```bash
npm install @valkey/valkey-glide
```

Here's the example code:

> ```javascript

Reviewer: validate code example again against glide api to make sure

                for (let x = start; x <= end; x++) {
                    batch[`foo${x}`] = x.toString();
                }
                
                // Execute batch mset
                await client.mset(batch);
                
                // Update counter and display progress
                await client.set("__last__", end.toString());
                console.log(`Batch completed: ${start} to ${end}`);
                
                // Verify a sample key from the batch
                const sampleKey = `foo${start}`;
                const value = await client.get(sampleKey);
                console.log(`Sample verification - ${sampleKey}: ${value}`);
                
            } catch (error) {
                console.log(`Error in batch starting at ${start}: ${error.message}`);
            }
        }
    } catch (error) {
        if (error instanceof ConnectionError) {
            console.log(`Connection error: ${error.message}`);
        } else {
            console.log(`Unexpected error: ${error.message}`);
        }
    } finally {
        if (client) {
            client.close();
        }
    }
}

runExample().catch(console.error);
```

> The application does a very simple thing, it sets keys in the form `foo<number>` to `number`, using batched MSET operations for better performance. So if you run the program the result is batches of

Reviewer: to make sure, validate the correctness of the text

is writing to the cluster.

This section explains some basic usage of
[Valkey GLIDE for Node.js](https://github.com/valkey-io/valkey-glide/tree/main/node), the official
Valkey client library, showing a simple example application.

The following example demonstrates how to connect to a Valkey cluster and perform
basic operations. First, install the Valkey GLIDE client:

```bash
npm install @valkey/valkey-glide
```

Here's the example code:

```javascript
import { GlideClusterClient, ConnectionError } from "@valkey/valkey-glide";

async function runExample() {
    // Define startup nodes - you only need one reachable node

>     let startupNodes = [

Reviewer: this is not the correct naming, we dont use startupNodes naming as it is not reflect the real esence of those, we use adresses, when there is config end point to fallback to in case of adress changes we advise to use it, but for generic, we use adressess as the name

        { host: "127.0.0.1", port: 7000 },
        { host: "127.0.0.1", port: 7001 }
    ];

    // Handle command line arguments

>     if (process.argv.length === 4) {

Reviewer: it is not fitted for nodejs to have code which get addresses from command line

        const [,, host, port] = process.argv;
        startupNodes = [{ host, port: parseInt(port) }];
    }

    let client;
    
    try {
        // Create cluster client
        client = await GlideClusterClient.createClient({

>             addresses: startupNodes,

Reviewer: addresses are part of the configuration which in case of cluster are GlideClusterClientConfiguration. validate naming and API for everything you write.

            clientConfiguration: {
                requestTimeout: 100
            }
        });

        console.log("Connected to Valkey cluster");

        // Get the last counter value, or start from 0
        let last = await client.get("__last__");
        last = last ? parseInt(last) : 0;

        console.log(`Starting from counter: ${last}`);

        // Write keys in sequence

>         for (let x = last + 1; x <= 1000000000; x++) {

Reviewer: Use batches of mset with logical amount of keys per batch, validate the api.

            try {
                await client.set(`foo${x}`, x.toString());
                const value = await client.get(`foo${x}`);
                console.log(value);
                await client.set("__last__", x.toString());
            } catch (error) {
                console.log(`Error: ${error.message}`);
            }
            
            // Sleep for 100ms to make output readable

>             await new Promise(resolve => setTimeout(resolve, 100));

Reviewer: why? which language convention you are following? doesnt make any sense

        }
    } catch (error) {
        if (error instanceof ConnectionError) {
            console.log(`Connection error: ${error.message}`);
        } else {
            console.log(`Unexpected error: ${error.message}`);
        }
    } finally {
        if (client) {
            client.close();
        }
    }
}

runExample().catch(console.error);
```

The application does a very simple thing, it sets keys in the form `foo<number>` to `number`, one after the other. So if you run the program the result is the
following stream of commands:

* SET foo0 0

> * SET foo1 1  

Reviewer: remove empty space

is writing to the cluster.

This section explains some basic usage of
[Valkey GLIDE for Node.js](https://github.com/valkey-io/valkey-glide/tree/main/node), an official

> Valkey client library, available in numerous languages, showing a practical example application in Node.js.

Reviewer 4: This is slightly confusing. GLIDE is 'available in numerous languages' not Valkey GLIDE for Node.js.

I might rephrase this entire passage.

> This section explains some basic usage of
> [Valkey GLIDE for Node.js](https://github.com/valkey-io/valkey-glide/tree/main/node), an official
> Valkey client library, available in numerous languages, showing a practical example application in Node.js.

Reviewer: ```suggestion
This section showcases the core functionality of Valkey through a practical Node.js application.  
For our example, we will use the [Node.js version of GLIDE,](https://github.com/valkey-io/valkey-glide/tree/main/node) an official Valkey client library that supports multiple languages.
```

runExample().catch(console.error);
```

> The application writes keys in the format `foo<number>` to `number`, one after the other, following the same logic as the original Ruby example. For each key, it performs a SET operation, immediately verifies the value with a GET operation, updates a counter, and includes a small delay between operations.

Reviewer: @copilot

written is the right thing.

From our point of view the cluster receiving the writes could just always
write the key `foo` to `42` to every operation, and we would not notice at
all.

Now we can write a more interesting application for testing cluster behavior.
A comprehensive consistency checking application that uses a set of counters, by default 1000, and sends `INCR` commands to increment the counters.

However instead of just writing, the application does two additional things:

* When a counter is updated using `INCR`, the application remembers the write.
* It also reads a random counter before every write, and check if the value is what we expected it to be, comparing it with the value it has in memory.

What this means is that this application is a **consistency checker**,
and is able to tell you if the cluster lost some write, or if it accepted
a write that we did not receive acknowledgment for. In the first case we'll
see a counter having a value that is smaller than the one we remember, while
in the second case the value will be greater.

Running a consistency testing application produces a line of output every
second:

> ```

Reviewer 4: tag this code block correctly.

Reviewer: please explain

Reviewer 4: Sorry, wrong line looks like it should be line 612

Adding a new replica can be performed in two ways. The obvious one is to
use valkey-cli again, but with the --cluster-replica option, like this:

    `valkey-cli --cluster add-node 127.0.0.1:7006 127.0.0.1:7000 --cluster-replica

> `

Reviewer 4: Have you tested this in a browser? I expect this might look weird since the backtick spans a line.

However you can specify exactly what primary you want to target with your
new replica with the following command line:

>     `valkey-cli --cluster add-node 127.0.0.1:7006 127.0.0.1:7000 --cluster-replica --cluster-master-id 3c3a0c74aae0b56170ccb03a76b60cfe7dc1912e

Reviewer 4: see above about testing in browser.

To remove a replica node just use the `del-node` command of valkey-cli:

    `valkey-cli --cluster del-node 127.0.0.1:7000 <node-id>`

> The first argument is just a random node in the cluster, the second argument

Reviewer: ```suggestion
    `valkey-cli --cluster del-node 127.0.0.1:7000 <node-id>`
```

> To remove a replica node just use the `del-node` command of valkey-cli:

Reviewer: ```suggestion
    `valkey-cli --cluster del-node 127.0.0.1:7000 <node-id>`
```

> To remove a replica node just use the `del-node` command of valkey-cli:

Reviewer: ```suggestion
    `valkey-cli --cluster del-node 127.0.0.1:7000 <node-id>`
```

>     `valkey-cli --cluster del-node 127.0.0.1:7000 `<node-id>``

Reviewer 4: looks like an extra backtick

Reviewer: `<node-id>` should i remove the backtick from this and leave the wrapping ones?

You should not use the `del-node` command because it tries to connect to all nodes and you will encounter a "connection refused" error.
Instead, you can use the `call` command:

>     `valkey-cli --cluster call 127.0.0.1:7000 cluster forget `<node-id>``

Reviewer 4: extra backtick

Reviewer: same q as above

Reviewer 4: In general, this looks like it hasn't been reviewed in a full preview.
